index/faults/faults.go

```diff
@@ -119,163 +118,120 @@ func (s *Index) Close() error {
+		log.Errorf("initial updating faults index: %s", err)
 	}
 	for {
 		select {
 		case <-s.ctx.Done():
-			log.Info("graceful shutdown of background slashing updater")
+			log.Info("graceful shutdown of background faults updater")
 			return
 		case <-time.After(util.AvgBlockTime):
 			if err := s.updateIndex(); err != nil {
-				log.Errorf("error when updating slashing history: %s", err)
+				log.Errorf("updating faults history: %s", err)
 				continue
 			}
 		}
 	}
 }
 
-// updateIndex updates current index with a new discovered chain head.
+// updateIndex updates current index.
 func (s *Index) updateIndex() error {
-	log.Info("updating slashing index...")
-	heaviest, err := s.api.ChainHead(s.ctx)
+	log.Info("updating faults index...")
+
+	chainHead, err := s.api.ChainHead(s.ctx)
 	if err != nil {
 		return fmt.Errorf("getting chain head: %s", err)
 	}
-	if heaviest.Height()-hOffset <= 0 {
+	// If the chain is very young, wait a bit for start building the index.
+	if chainHead.Height()-hOffset <= 0 {
 		return nil
 	}
-	new, err := s.api.ChainGetTipSetByHeight(s.ctx, heaviest.Height()-hOffset, heaviest.Key())
+
+	// Get the tipset hOffset before current head as the target tipset
+	// to let the index be built.
+	targetTs, err := s.api.ChainGetTipSetByHeight(s.ctx, chainHead.Height()-hOffset, chainHead.Key())
 	if err != nil {
-		return fmt.Errorf("get tipset by height: %s", err)
+		return fmt.Errorf("getting offseted tipset from head: %s", err)
 	}
-	newtsk := types.NewTipSetKey(new.Cids()...)
+
 	var index IndexSnapshot
-	ts, err := s.store.LoadAndPrune(s.ctx, newtsk, &index)
+	// Load the last saved index snapshot which is a child of the target TipSet.
+	// Might not be s.index because of reorgs.
+	indexTs, err := s.store.LoadAndPrune(s.ctx, targetTs.Key(), &index)
 	if err != nil {
 		return fmt.Errorf("load tipset state: %s", err)
 	}
 	if index.Miners == nil {
-		index.Miners = make(map[string]Slashes)
+		index.Miners = make(map[string]Faults)
 	}
-	_, path, err := chainsync.ResolveBase(s.ctx, s.api, ts, newtsk)
+
+	// Get the tipset path between the indexTs and the targetTs, so to
+	// calculate the faults that happened between last saved index and target.
+	_, path, err := chainsync.ResolveBase(s.ctx, s.api, indexTs, targetTs.Key())
 	if err != nil {
 		return fmt.Errorf("resolving base path: %s", err)
 	}
+
 	mctx := context.Background()
 	start := time.Now()
 	for i := 0; i < len(path); i += batchSize {
 		j := i + batchSize
 		if j > len(path) {
 			j = len(path)
 		}
-		if err := updateFromPath(s.ctx, s.api, &index, path[i:j]); err != nil {
-			return fmt.Errorf("getting update from path section: %s", err)
+
+		// Get section head and length, and get all faults happened
+		// in this section and include it into the updating index.
+		sectionLength := abi.ChainEpoch(j - i)
+		sectionHeadTs := path[j-1].Key()
+		faults, err := s.api.StateAllMinerFaults(s.ctx, sectionLength, sectionHeadTs)
```

Comment:
So this is the new API we're leveraging to get miners faults.
We abandoned a more complex strategy getting internal miner state and inspecting fields.